cpp/test/sparse/solver/lanczos.cu

```diff
@@ -0,0 +1,434 @@
+      v0(raft::make_device_vector<ValueType, uint32_t, raft::row_major>(handle, n)),
+      eigenvalues(raft::make_device_vector<ValueType, uint32_t, raft::col_major>(
+        handle, params.n_components)),
+      eigenvectors(raft::make_device_matrix<ValueType, uint32_t, raft::col_major>(
+        handle, n, params.n_components)),
+      expected_eigenvalues(
+        raft::make_device_vector<ValueType, uint32_t, raft::col_major>(handle, params.n_components))
+  {
+  }
+
+ protected:
+  void SetUp() override
+  {
+    raft::copy(rows.data_handle(), params.rows.data(), n + 1, stream);
+    raft::copy(cols.data_handle(), params.cols.data(), nnz, stream);
+    raft::copy(vals.data_handle(), params.vals.data(), nnz, stream);
+    raft::copy(expected_eigenvalues.data_handle(),
+               params.expected_eigenvalues.data(),
+               params.n_components,
+               stream);
+  }
+
+  void TearDown() override {}
+
+  void Run()
+  {
+    raft::random::uniform<ValueType>(handle, rng, v0.view(), 0, 1);
+    std::tuple<IndexType, ValueType, IndexType> stats;
+
+    raft::sparse::solver::lanczos_solver_config<ValueType> config{
+      params.n_components, params.maxiter, params.restartiter, params.tol, rng.seed};
+    auto csr_structure =
+      raft::make_device_compressed_structure_view<IndexType, IndexType, IndexType>(
+        const_cast<IndexType*>(rows.data_handle()),
+        const_cast<IndexType*>(cols.data_handle()),
+        n,
+        n,
+        nnz);
+
+    auto csr_matrix = raft::make_device_csr_matrix_view<ValueType, IndexType, IndexType, IndexType>(
+      const_cast<ValueType*>(vals.data_handle()), csr_structure);
+
+    std::get<0>(stats) =
+      raft::sparse::solver::lanczos_compute_smallest_eigenvectors<IndexType, ValueType>(
+        handle, csr_matrix, config, v0.view(), eigenvalues.view(), eigenvectors.view());
+
+    ASSERT_TRUE(raft::devArrMatch<ValueType>(eigenvalues.data_handle(),
+                                             expected_eigenvalues.data_handle(),
+                                             params.n_components,
+                                             raft::CompareApprox<ValueType>(1e-5),
+                                             stream));
+  }
+
+ protected:
+  lanczos_inputs<IndexType, ValueType> params;
+  raft::resources handle;
+  cudaStream_t stream;
+  int n;
+  int nnz;
+  raft::random::RngState rng;
+  raft::device_vector<IndexType, uint32_t, raft::row_major> rows;
+  raft::device_vector<IndexType, uint32_t, raft::row_major> cols;
+  raft::device_vector<ValueType, uint32_t, raft::row_major> vals;
+  raft::device_vector<ValueType, uint32_t, raft::row_major> v0;
+  raft::device_vector<ValueType, uint32_t, raft::col_major> eigenvalues;
+  raft::device_matrix<ValueType, uint32_t, raft::col_major> eigenvectors;
+  raft::device_vector<ValueType, uint32_t, raft::col_major> expected_eigenvalues;
+};
+
+const std::vector<lanczos_inputs<int, float>> inputsf = {
+  {2,
```

Comment:
Please add to the ongoing follow-on issue to find a way to generate and validate test data without hardcoding. I understand why Anupam did this, and I agree that it's not straightforward to find a way to automate this in the tests, but we need to expend some time at some point to derive a good naive validation.